diff --git a/CHANGELOG.md b/CHANGELOG.md
deleted file mode 100644
index 588329cd3d515134aab5a5967298b09a0faf1a5c..0000000000000000000000000000000000000000
diff --git a/dist/index.d.cts b/dist/index.d.cts
index 79b1ccf28444fd9828a7fde4cf27b542b1d62487..fab7c940b5472efd07bff8d1ee51678350969ac4 100644
--- a/dist/index.d.cts
+++ b/dist/index.d.cts
@@ -26,31 +26,34 @@ type IsForm<T$1> = T$1 extends string ? T$1 extends `multipart/form-data${infer
 type ReturnJsonOrTextOrResponse<ContentType, Content$1, Status$1 extends keyof StatusCodeRangeDefinitions | StatusCode> = ContentType extends string ? ContentType extends `application/${infer Start}json${infer _End}` ? Start extends "" | `${string}+` | `vnd.${string}+` ? TypedResponse<JSONParsed<Content$1>, ExtractStatusCode<Status$1>, "json"> : never : ContentType extends `text/plain${infer _Rest}` ? TypedResponse<Content$1, ExtractStatusCode<Status$1>, "text"> : Response : never;
 type RequestPart<R$1 extends RouteConfig, Part extends string> = Part extends keyof R$1["request"] ? R$1["request"][Part] : {};
 type HasUndefined<T$1> = undefined extends T$1 ? true : false;
+type InferSchemaInput<T$1> = T$1 extends { _input: infer Input } ? Input : never;
+type InferSchemaOutput<T$1> = T$1 extends { _output: infer Output } ? Output : never;
+type InferSchemaContent<T$1> = T$1 extends { _output: infer Output } ? Output : never;
 type InputTypeBase<R$1 extends RouteConfig, Part extends string, Type extends keyof ValidationTargets> = R$1["request"] extends RequestTypes ? RequestPart<R$1, Part> extends ZodType ? {
   in: { [K in Type]: HasUndefined<ValidationTargets[K]> extends true ? { [K2 in keyof z.input<RequestPart<R$1, Part>>]?: z.input<RequestPart<R$1, Part>>[K2] } : { [K2 in keyof z.input<RequestPart<R$1, Part>>]: z.input<RequestPart<R$1, Part>>[K2] } };
   out: { [K in Type]: z.output<RequestPart<R$1, Part>> };
 } : {} : {};
-type InputTypeJson<R$1 extends RouteConfig> = R$1["request"] extends RequestTypes ? R$1["request"]["body"] extends ZodRequestBody ? R$1["request"]["body"]["content"] extends ZodContentObject ? IsJson<keyof R$1["request"]["body"]["content"]> extends never ? {} : R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]] extends Record<"schema", ZodType<any>> ? {
+type InputTypeJson<R$1 extends RouteConfig> = R$1["request"] extends RequestTypes ? R$1["request"]["body"] extends ZodRequestBody ? R$1["request"]["body"]["content"] extends ZodContentObject ? IsJson<keyof R$1["request"]["body"]["content"]> extends never ? {} : R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]] extends Record<"schema", infer Schema> ? {
   in: {
-    json: z.input<R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]]["schema"]>;
+    json: InferSchemaInput<Schema>;
   };
   out: {
-    json: z.output<R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]]["schema"]>;
+    json: InferSchemaOutput<Schema>;
   };
 } : {} : {} : {} : {};
-type InputTypeForm<R$1 extends RouteConfig> = R$1["request"] extends RequestTypes ? R$1["request"]["body"] extends ZodRequestBody ? R$1["request"]["body"]["content"] extends ZodContentObject ? IsForm<keyof R$1["request"]["body"]["content"]> extends never ? {} : R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]] extends Record<"schema", ZodType<any>> ? {
+type InputTypeForm<R$1 extends RouteConfig> = R$1["request"] extends RequestTypes ? R$1["request"]["body"] extends ZodRequestBody ? R$1["request"]["body"]["content"] extends ZodContentObject ? IsForm<keyof R$1["request"]["body"]["content"]> extends never ? {} : R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]] extends Record<"schema", infer Schema> ? {
   in: {
-    form: z.input<R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]]["schema"]>;
+    form: InferSchemaInput<Schema>;
   };
   out: {
-    form: z.output<R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]]["schema"]>;
+    form: InferSchemaOutput<Schema>;
   };
 } : {} : {} : {} : {};
 type InputTypeParam<R$1 extends RouteConfig> = InputTypeBase<R$1, "params", "param">;
 type InputTypeQuery<R$1 extends RouteConfig> = InputTypeBase<R$1, "query", "query">;
 type InputTypeHeader<R$1 extends RouteConfig> = InputTypeBase<R$1, "headers", "header">;
 type InputTypeCookie<R$1 extends RouteConfig> = InputTypeBase<R$1, "cookies", "cookie">;
-type ExtractContent<T$1> = T$1 extends { [K in keyof T$1]: infer A } ? A extends Record<"schema", ZodType> ? z.infer<A["schema"]> : never : never;
+type ExtractContent<T$1> = T$1 extends { [K in keyof T$1]: infer A } ? A extends Record<"schema", infer Schema> ? InferSchemaContent<Schema> : never : never;
 type StatusCodeRangeDefinitions = {
   "1XX": InfoStatusCode;
   "2XX": SuccessStatusCode;
@@ -177,11 +180,11 @@ declare class OpenAPIHono<E$1 extends Env = Env, S$1 extends Schema = {}, BasePa
   *  }
   *)
   */
-  openapi: <R extends RouteConfig, I extends Input = InputTypeParam<R> & InputTypeQuery<R> & InputTypeHeader<R> & InputTypeCookie<R> & InputTypeForm<R> & InputTypeJson<R>, P extends string = ConvertPathType<R["path"]>>({
+  openapi: <const R extends RouteConfig, P extends string = ConvertPathType<R["path"]>>({
     middleware: routeMiddleware,
     hide,
     ...route
-  }: R, handler: Handler<R["middleware"] extends MiddlewareHandler[] | MiddlewareHandler ? RouteMiddlewareParams<R>["env"] & E$1 : E$1, P, I, R extends {
+  }: R, handler: Handler<R["middleware"] extends MiddlewareHandler[] | MiddlewareHandler ? RouteMiddlewareParams<R>["env"] & E$1 : E$1, P, InputTypeParam<R> & InputTypeQuery<R> & InputTypeHeader<R> & InputTypeCookie<R> & InputTypeForm<R> & InputTypeJson<R>, R extends {
     responses: {
       [statusCode: number]: {
         content: {
@@ -189,7 +192,7 @@ declare class OpenAPIHono<E$1 extends Env = Env, S$1 extends Schema = {}, BasePa
         };
       };
     };
-  } ? MaybePromise<RouteConfigToTypedResponse<R>> : MaybePromise<RouteConfigToTypedResponse<R>> | MaybePromise<Response>>, hook?: Hook<I, E$1, P, R extends {
+  } ? MaybePromise<RouteConfigToTypedResponse<R>> : MaybePromise<RouteConfigToTypedResponse<R>> | MaybePromise<Response>>, hook?: Hook<InputTypeParam<R> & InputTypeQuery<R> & InputTypeHeader<R> & InputTypeCookie<R> & InputTypeForm<R> & InputTypeJson<R>, E$1, P, R extends {
     responses: {
       [statusCode: number]: {
         content: {
@@ -197,7 +200,7 @@ declare class OpenAPIHono<E$1 extends Env = Env, S$1 extends Schema = {}, BasePa
         };
       };
     };
-  } ? MaybePromise<RouteConfigToTypedResponse<R>> | undefined : MaybePromise<RouteConfigToTypedResponse<R>> | MaybePromise<Response> | undefined> | undefined) => OpenAPIHono<E$1, S$1 & ToSchema<R["method"], MergePath<BasePath, P>, I, RouteConfigToTypedResponse<R>>, BasePath>;
+  } ? MaybePromise<RouteConfigToTypedResponse<R>> | undefined : MaybePromise<RouteConfigToTypedResponse<R>> | MaybePromise<Response> | undefined> | undefined) => OpenAPIHono<E$1, S$1 & ToSchema<R["method"], MergePath<BasePath, P>, InputTypeParam<R> & InputTypeQuery<R> & InputTypeHeader<R> & InputTypeCookie<R> & InputTypeForm<R> & InputTypeJson<R>, RouteConfigToTypedResponse<R>>, BasePath>;
   getOpenAPIDocument: (objectConfig: OpenAPIObjectConfig, generatorConfig?: OpenAPIGeneratorOptions) => OpenAPIObject;
   getOpenAPI31Document: (objectConfig: OpenAPIObjectConfig, generatorConfig?: OpenAPIGeneratorOptions) => OpenAPIObject$1;
   doc: <P extends string>(path: P, configureObject: OpenAPIObjectConfigure<E$1, P>, configureGenerator?: OpenAPIGeneratorConfigure<E$1, P>) => OpenAPIHono<E$1, S$1 & ToSchema<"get", MergePath<BasePath, P>, {}, {}>, BasePath>;
@@ -209,7 +212,7 @@ declare class OpenAPIHono<E$1 extends Env = Env, S$1 extends Schema = {}, BasePa
   notFound: (handler: NotFoundHandler<E$1>) => OpenAPIHono<E$1, S$1, BasePath>;
 }
 type RoutingPath<P$1 extends string> = P$1 extends `${infer Head}/{${infer Param}}${infer Tail}` ? `${Head}/:${Param}${RoutingPath<Tail>}` : P$1;
-declare const createRoute: <P extends string, R extends Omit<RouteConfig, "path"> & {
+declare const createRoute: <P extends string, const R extends Omit<RouteConfig, "path"> & {
   path: P;
 }>(routeConfig: R) => R & {
   getRoutingPath(): RoutingPath<R["path"]>;
diff --git a/dist/index.d.ts b/dist/index.d.ts
index 7355946a2d26b4626bd89f7eaf03d0351f9d9d33..1092dece6e5201061d6b52e23aa71325533b6b76 100644
--- a/dist/index.d.ts
+++ b/dist/index.d.ts
@@ -26,31 +26,34 @@ type IsForm<T$1> = T$1 extends string ? T$1 extends `multipart/form-data${infer
 type ReturnJsonOrTextOrResponse<ContentType, Content$1, Status$1 extends keyof StatusCodeRangeDefinitions | StatusCode> = ContentType extends string ? ContentType extends `application/${infer Start}json${infer _End}` ? Start extends "" | `${string}+` | `vnd.${string}+` ? TypedResponse<JSONParsed<Content$1>, ExtractStatusCode<Status$1>, "json"> : never : ContentType extends `text/plain${infer _Rest}` ? TypedResponse<Content$1, ExtractStatusCode<Status$1>, "text"> : Response : never;
 type RequestPart<R$1 extends RouteConfig, Part extends string> = Part extends keyof R$1["request"] ? R$1["request"][Part] : {};
 type HasUndefined<T$1> = undefined extends T$1 ? true : false;
+type InferSchemaInput<T$1> = T$1 extends { _input: infer Input } ? Input : never;
+type InferSchemaOutput<T$1> = T$1 extends { _output: infer Output } ? Output : never;
+type InferSchemaContent<T$1> = T$1 extends { _output: infer Output } ? Output : never;
 type InputTypeBase<R$1 extends RouteConfig, Part extends string, Type extends keyof ValidationTargets> = R$1["request"] extends RequestTypes ? RequestPart<R$1, Part> extends ZodType ? {
   in: { [K in Type]: HasUndefined<ValidationTargets[K]> extends true ? { [K2 in keyof z.input<RequestPart<R$1, Part>>]?: z.input<RequestPart<R$1, Part>>[K2] } : { [K2 in keyof z.input<RequestPart<R$1, Part>>]: z.input<RequestPart<R$1, Part>>[K2] } };
   out: { [K in Type]: z.output<RequestPart<R$1, Part>> };
 } : {} : {};
-type InputTypeJson<R$1 extends RouteConfig> = R$1["request"] extends RequestTypes ? R$1["request"]["body"] extends ZodRequestBody ? R$1["request"]["body"]["content"] extends ZodContentObject ? IsJson<keyof R$1["request"]["body"]["content"]> extends never ? {} : R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]] extends Record<"schema", ZodType<any>> ? {
+type InputTypeJson<R$1 extends RouteConfig> = R$1["request"] extends RequestTypes ? R$1["request"]["body"] extends ZodRequestBody ? R$1["request"]["body"]["content"] extends ZodContentObject ? IsJson<keyof R$1["request"]["body"]["content"]> extends never ? {} : R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]] extends Record<"schema", infer Schema> ? {
   in: {
-    json: z.input<R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]]["schema"]>;
+    json: InferSchemaInput<Schema>;
   };
   out: {
-    json: z.output<R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]]["schema"]>;
+    json: InferSchemaOutput<Schema>;
   };
 } : {} : {} : {} : {};
-type InputTypeForm<R$1 extends RouteConfig> = R$1["request"] extends RequestTypes ? R$1["request"]["body"] extends ZodRequestBody ? R$1["request"]["body"]["content"] extends ZodContentObject ? IsForm<keyof R$1["request"]["body"]["content"]> extends never ? {} : R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]] extends Record<"schema", ZodType<any>> ? {
+type InputTypeForm<R$1 extends RouteConfig> = R$1["request"] extends RequestTypes ? R$1["request"]["body"] extends ZodRequestBody ? R$1["request"]["body"]["content"] extends ZodContentObject ? IsForm<keyof R$1["request"]["body"]["content"]> extends never ? {} : R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]] extends Record<"schema", infer Schema> ? {
   in: {
-    form: z.input<R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]]["schema"]>;
+    form: InferSchemaInput<Schema>;
   };
   out: {
-    form: z.output<R$1["request"]["body"]["content"][keyof R$1["request"]["body"]["content"]]["schema"]>;
+    form: InferSchemaOutput<Schema>;
   };
 } : {} : {} : {} : {};
 type InputTypeParam<R$1 extends RouteConfig> = InputTypeBase<R$1, "params", "param">;
 type InputTypeQuery<R$1 extends RouteConfig> = InputTypeBase<R$1, "query", "query">;
 type InputTypeHeader<R$1 extends RouteConfig> = InputTypeBase<R$1, "headers", "header">;
 type InputTypeCookie<R$1 extends RouteConfig> = InputTypeBase<R$1, "cookies", "cookie">;
-type ExtractContent<T$1> = T$1 extends { [K in keyof T$1]: infer A } ? A extends Record<"schema", ZodType> ? z.infer<A["schema"]> : never : never;
+type ExtractContent<T$1> = T$1 extends { [K in keyof T$1]: infer A } ? A extends Record<"schema", infer Schema> ? InferSchemaContent<Schema> : never : never;
 type StatusCodeRangeDefinitions = {
   "1XX": InfoStatusCode;
   "2XX": SuccessStatusCode;
@@ -177,11 +180,11 @@ declare class OpenAPIHono<E$1 extends Env = Env, S$1 extends Schema = {}, BasePa
   *  }
   *)
   */
-  openapi: <R extends RouteConfig, I extends Input = InputTypeParam<R> & InputTypeQuery<R> & InputTypeHeader<R> & InputTypeCookie<R> & InputTypeForm<R> & InputTypeJson<R>, P extends string = ConvertPathType<R["path"]>>({
+  openapi: <const R extends RouteConfig, P extends string = ConvertPathType<R["path"]>>({
     middleware: routeMiddleware,
     hide,
     ...route
-  }: R, handler: Handler<R["middleware"] extends MiddlewareHandler[] | MiddlewareHandler ? RouteMiddlewareParams<R>["env"] & E$1 : E$1, P, I, R extends {
+  }: R, handler: Handler<R["middleware"] extends MiddlewareHandler[] | MiddlewareHandler ? RouteMiddlewareParams<R>["env"] & E$1 : E$1, P, InputTypeParam<R> & InputTypeQuery<R> & InputTypeHeader<R> & InputTypeCookie<R> & InputTypeForm<R> & InputTypeJson<R>, R extends {
     responses: {
       [statusCode: number]: {
         content: {
@@ -189,7 +192,7 @@ declare class OpenAPIHono<E$1 extends Env = Env, S$1 extends Schema = {}, BasePa
         };
       };
     };
-  } ? MaybePromise<RouteConfigToTypedResponse<R>> : MaybePromise<RouteConfigToTypedResponse<R>> | MaybePromise<Response>>, hook?: Hook<I, E$1, P, R extends {
+  } ? MaybePromise<RouteConfigToTypedResponse<R>> : MaybePromise<RouteConfigToTypedResponse<R>> | MaybePromise<Response>>, hook?: Hook<InputTypeParam<R> & InputTypeQuery<R> & InputTypeHeader<R> & InputTypeCookie<R> & InputTypeForm<R> & InputTypeJson<R>, E$1, P, R extends {
     responses: {
       [statusCode: number]: {
         content: {
@@ -197,7 +200,7 @@ declare class OpenAPIHono<E$1 extends Env = Env, S$1 extends Schema = {}, BasePa
         };
       };
     };
-  } ? MaybePromise<RouteConfigToTypedResponse<R>> | undefined : MaybePromise<RouteConfigToTypedResponse<R>> | MaybePromise<Response> | undefined> | undefined) => OpenAPIHono<E$1, S$1 & ToSchema<R["method"], MergePath<BasePath, P>, I, RouteConfigToTypedResponse<R>>, BasePath>;
+  } ? MaybePromise<RouteConfigToTypedResponse<R>> | undefined : MaybePromise<RouteConfigToTypedResponse<R>> | MaybePromise<Response> | undefined> | undefined) => OpenAPIHono<E$1, S$1 & ToSchema<R["method"], MergePath<BasePath, P>, InputTypeParam<R> & InputTypeQuery<R> & InputTypeHeader<R> & InputTypeCookie<R> & InputTypeForm<R> & InputTypeJson<R>, RouteConfigToTypedResponse<R>>, BasePath>;
   getOpenAPIDocument: (objectConfig: OpenAPIObjectConfig, generatorConfig?: OpenAPIGeneratorOptions) => OpenAPIObject;
   getOpenAPI31Document: (objectConfig: OpenAPIObjectConfig, generatorConfig?: OpenAPIGeneratorOptions) => OpenAPIObject$1;
   doc: <P extends string>(path: P, configureObject: OpenAPIObjectConfigure<E$1, P>, configureGenerator?: OpenAPIGeneratorConfigure<E$1, P>) => OpenAPIHono<E$1, S$1 & ToSchema<"get", MergePath<BasePath, P>, {}, {}>, BasePath>;
@@ -209,7 +212,7 @@ declare class OpenAPIHono<E$1 extends Env = Env, S$1 extends Schema = {}, BasePa
   notFound: (handler: NotFoundHandler<E$1>) => OpenAPIHono<E$1, S$1, BasePath>;
 }
 type RoutingPath<P$1 extends string> = P$1 extends `${infer Head}/{${infer Param}}${infer Tail}` ? `${Head}/:${Param}${RoutingPath<Tail>}` : P$1;
-declare const createRoute: <P extends string, R extends Omit<RouteConfig, "path"> & {
+declare const createRoute: <P extends string, const R extends Omit<RouteConfig, "path"> & {
   path: P;
 }>(routeConfig: R) => R & {
   getRoutingPath(): RoutingPath<R["path"]>;
